kitsune/sumo/middleware.py

```diff
@@ -108,76 +105,113 @@ def process_request(self, request):
                 logout(request)
 
 
-class LocaleURLMiddleware(MiddlewareMixin):
+class LocaleMiddleware:
```

Comment:
This is a great question, one I thought a lot about while working on this PR, and one that @smithellis and I talked a long time last week too. We had wished that you were there, so now we'll get a chance to run an idea by you. First, I'll answer your numbered questions:

> 1. If we use the old style middleware, do we avoid subclassing the process_response?

No, because we'd still need to check if the response is a redirect, and adjust the locale of the redirect, if needed. This is because Django's `LocaleMiddleware` takes care of redirecting a requested URL, without a language in the path, to the same URL prefixed with a language, if that original URL was a 404, and the new, language-prefixed URL would resolve to a valid path.

> 2. Can we have a custom sumo middleware that will be right after Django's default in the middleware order responsible for applying our custom logic?

I originally developed the `kitsune.sumo.middleware.LocaleMiddleware` in this PR as a sub-class of Django's `LocaleMiddleware` (or it could have been a separate middleware placed right before Django's in the middleware chain), but realized later I couldn't do that due to SUMO's custom approach for getting the language from the request when there's no language in the path. Django does it this way:

1. Try getting the language from path.
2. If no supported language was found, try getting the language from the Django language cookie.
3. If no supported language was found, try getting the language from the `Accept-Language` header.
4. If no supported language was found, default to `settings.LANGUAGE_CODE`.

While SUMO currently does it this way:

1. Try getting the language from path.
2. If no supported language was found, try getting it from the user's profile if the user is logged in.
3. If no supported language was found, try getting it from the session if the user is not logged in. The session would have a language stored only if a URL with the `lang` parameter had been requested previously in the session.
4. If no supported language was found, try getting the language from the Django language cookie.
5. If no supported language was found, try getting the language from the `Accept-Language` header.
6. If no supported language was found, default to `settings.LANGUAGE_CODE`.

So the difference is SUMO's special logic in steps `2` and `3`.

But after discussing this with @smithellis and doing some thinking, I might be able to sub-class Django's `LocaleMiddleware` (or create a separate one that we place right before Django's) if we make some changes to how we implement SUMO's steps `2` and `3`.

Reply:
I just realized something today that for me kills any hope of finding a nice way to create a SUMO locale middleware that sub-classes or works with Django's `LocaleMiddleware. As mentioned earlier, Django's `LocaleMiddleware` takes care of redirecting a requested URL without a language in the path, to the same URL prefixed with a language, if that original URL was a 404, and the new, language-prefixed URL would resolve to a valid path. That code doesn't work for us, because it prefixes a language code with a lowercase country-code (e.g., `en-us`) and then only redirects if the new prefixed URL is a valid path (it calls Django's `is_valid_path`). However, any language code with a lowercase country-code is never a valid path, so the redirects are never handled for those cases.

In the end, SUMO's differences (uppercase country-codes, non-standard language fallbacks like `sc` to `it`, checking for the locale in the user's profile and session, etc.) just make it impossible to create something that plays well with Django's `LocaleMiddleware`. I think we just have to swallow that we'll have to maintain our own `LocaleMiddleware` with some differences and lots of similarities to Django's.

Reply:
Can't we just inherit from Django's Localmiddleware, call super() and adjust the response? Maybe we can discuss this more through slack

Reply:
I've added a comment on this in the main thread below.